adding `sub_saveat` functionality to `simulate()` for better memory performance.

Nick-P-Adams opened the pull request:
- Changed `odes.simulate()` added `sub_save_indicies` and `save_step` optional parameters that are passed to `build_saveat` function.
- Added `odes.build_saveat()` to determine (`sub_save_indicies is not None`) if `SubSaveAt` object should be used when building the `SaveAt` object. `build_saveat()` can also optionally increment the time steps `SaveAt` saves states via the `save_step` argument.

arik-shurygin commented on examples/sir.py:
```diff
@@ -97,7 +101,27 @@ def run_simulation(config: SimulationConfig, tf) -> Solution:
     return solution
 
 
+def run_simulation_sub_save_example(config: SimulationConfig, tf) -> Solution:
```
I think that we should aim to keep `sir.py` as clean as possible, so maybe we take this example and try to use it in an `examples` section of the `simulate` docstring. This way we can still show off how to use `sub_save_indices` and `save_step` but without adding to the confusion of a presumably new user.

arik-shurygin commented on examples/sir.py:
```diff
@@ -121,26 +145,47 @@ def model(
 plt.legend()
 plt.show()
+
+# example of using sub save for compartments
```
I think same point about keeping sir clean applies here, especially since the plot will be basically the same but with lower fidelity and less compartments.

arik-shurygin commented on src/dynode/odes.py:
```diff
@@ -37,6 +38,8 @@ def simulate(
     initial_state: CompartmentState,
     ode_parameters: AbstractODEParams,
     solver_parameters: SolverParams,
+    sub_save_indices: Optional[Tuple[int, ...]] = None,
```
maybe we add the example from `sir.py` into a section in this docstring?

arik-shurygin reviewed (changes requested):
some small requests, mostly aiming to keep `sir.py` as small and clear as possible for new users. Also could you add a version increment to `pyproject.toml` and a section in the `CHANGELOG.md` for this added functionality. You can copy paste the changelog section into your PR description so the git log is clear too

Nick-P-Adams commented:
As requested, I removed what was added to `sir.py` and instead added a basic example to the `odes.simulate()` docstring. Also incremented the `pyproject.toml` version and added an entry to the `CHANGELOG.md`.

arik-shurygin reviewed (approved):
LGTM!
